Fixed ordering of search indexes

ErisDS opened the pull request:
ref https://linear.app/ghost/issue/PROD-2106

- When rebasing the new 6.x branch on top of these changes we started getting failures due to content length
- This makes some sense, but what was odd was that we couldn't see exactly what was changing
- This was due to the matchers in the tests being overly specific and using anyString.
- The intention of snapshot testing is to leave as many values in their natural form as possible, and only use matchers for values that will reasonably change: IDs, dates, versions etc
  - Note: I have seen Claude implement snapshot testing this way and also nearly committed it cos it looks right at first glance!
- Once the matchers were modified to minimal versions, it became clear that the API result wasn't deterministic. By that I mean it didn't repeatedly return the exact same response - the order kept changing
- This is because we don't store milliseconds in our DB, we were only ordering by updated_at, and our tests create tag fixtures way too fast
- So there are 3 separate fixes in this changeset
   - Added "id DESC" to all the sort orders, so we get a deterministic order if updated_at is the same
   - Changed our fixtures to add a second to the date used for creating tags, so each tag is 1 second apart, else the tags are still in a nondeterministic order
   - Updated snapshot matchers as mentioned above

ErisDS committed:
Fixed ordering of search indexes

- When rebasing the new 6.x branch on top of these changes
  we started getting failures due to content length
- This makes some sense, but what was odd was that we couldn't
  see exactly what was changing
- This was due to the matchers in the tests being overly specific
  and using anyString.
- The intention of snapshot testing is to leave as many values
  in their natural form as possible, and only use matchers for
  values that will reasonably change: IDs, dates, versions etc
  - Note: I have seen Claude implement snapshot testing this way
    and also nearly committed it cos it looks right at first
    glance!
- Once the matchers were modified to minimal versions, it became
  clear that the API result wasn't deterministic. By that I mean
  it didn't repeatedly return the exact same response - the order
  kept changing
- This is because we don't store milliseconds in our DB, we were
  only ordering by updated_at, and our tests create tag fixtures
  way too fast
- So there are 3 separate fixes in this changeset
   - Added "id DESC" to all the sort orders, so we get a deterministic
     order if updated_at is the same
   - Changed our fixtures to add a second to the date used for
     creating tags, so each tag is 1 second apart, else
     the tags are still in a nondeterministic order
   - Updated snapshot matchers as mentioned above
- Note: we have to use anyString for user ids, rather than
  anyObjectId because the owner user id is 1
  THIS IS FIXED IN 6.x!!!! Yay! 🎉

ErisDS committed:
Separate tag creation

ErisDS committed:
snapshots

sagzy committed:
Used a bigger offset before creating tag fixtures, so that the default tag getting-started tag has an older updated_at in CI

sagzy committed:
Tried an even bigger one (1 minute), to see if we need a different fix

sagzy commented:
<!-- linear:replyTo:3149756776 -->

Yes! Still on my todo list, hoping to get to it finally this week

sagzy committed:
Used anyObjectId over anyString for User IDs

- in Ghost 6.0, we no longer have a hardcoded ID=1 for the owner user and have an object ID for all users

sagzy committed:
Revert "Tried an even bigger one (1 minute), to see if we need a different fix"

This reverts commit 315988c79c70838f73d1eb5fdc42e9728a8bf3e1.

sagzy committed:
Revert "Used a bigger offset before creating tag fixtures, so that the default tag getting-started tag has an older updated_at in CI"

This reverts commit 22b7c622690d6c7cdca4a067f0ecd4694ee6f95f.

sagzy committed:
Fixed linting
